util/tracing: detect use-after-Finish in all exported Span methods

andreimatei opened the pull request:
Most exported Span methods already have protection for being called
after the span was Finish()ed. This patch adds the protection to the
ones that were still missing it.

Release note: None

andreimatei committed:
util/tracing: small debugging enhancement

If configured to do so, capture a span's finish stack before marking the
span as finished, not after. If we do it after, that races with someone
else finding the span finished and trying to read the stack.

Release note: None

andreimatei committed:
util/tracing: detect use-after-Finish in all exported Span methods

Most exported Span methods already have protection for being called
after the span was Finish()ed. This patch adds the protection to the
ones that were still missing it.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/75424)
<!-- Reviewable:end -->

andreimatei committed:
util/log: improve a test

This test was calling log.ExpensiveLogEnabled(ctx) after the Span in the
ctx had been Finish()ed. That's a span use-after-Finish, no longer
allowed.

Release note: None

andreimatei commented:
The first commit is #76388; ignore.

abarganier reviewed (commented):
The changes <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, just one question around new potential places to trigger panics from.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75424#-:-Mvds9Gq3p8kH-QjG8ed:b-disd73)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @abarganier and @andreimatei)

___
*[pkg/util/tracing/span.go, line 228 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/75424#-MvdsFfj37ju0UNJk8bb:-MvdsFfj37ju0UNJk8bc:b-u8rzco) ([raw file](https://github.com/cockroachdb/cockroach/blob/c714c039f76efc754396e0e356edb3490702b5a9/pkg/util/tracing/span.go#L228)):*
> ```Go
> // Tracer exports the tracer this span was created using.
> func (sp *Span) Tracer() *Tracer {
> 	sp.detectUseAfterFinish()
> ```

Just curious, have we run anything such as roachtests to see if these new calls to `detectUseAfterFinish()` cause any new panics we didn't expect?



_Code quote:_
```Go
	sp.detectUseAfterFinish()
```


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75424)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @abarganier)

___
*[pkg/util/tracing/span.go, line 228 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/75424#-MvdsFfj37ju0UNJk8bb:-MveyOUUBWUNdF1vBZvR:b-8gauqs) ([raw file](https://github.com/cockroachdb/cockroach/blob/c714c039f76efc754396e0e356edb3490702b5a9/pkg/util/tracing/span.go#L228)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

Just curious, have we run anything such as roachtests to see if these new calls to `detectUseAfterFinish()` cause any new panics we didn't expect?
</blockquote></details>

I've kicked off a roachtest run now with this change plus the change to turn the span creation on in production (i.e. also in roachtests).


<!-- Sent from Reviewable.io -->

abarganier reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75424#-:-MvsuJLABOd37fBNSNjW:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @abarganier)



<!-- Sent from Reviewable.io -->

andreimatei commented:
The roachtest run did find one more problem, now fixed in https://github.com/cockroachdb/cockroach/pull/76534. So that was a good call. I'll continue to keep an eye out for more.

TFTR!

bors +

andreimatei commented:
bors r+
